src/debug/debugService.ts

```diff
@@ -0,0 +1,251 @@
+                p.report({ message: "Creating port-forwarding..."});
+                kubeChannel.showOutput(`Starting to create port-forward on pod ${podName}...`, "port-forward");
+                const proxyResult = await this.createPortForward(podName, portInfo.debug, portInfo.app);
+                kubeChannel.showOutput(`Finished to create port-forward ${proxyResult.proxyDebugPort}:${portInfo.debug} ${proxyResult.proxyAppPort}:${portInfo.app}`);
+
+                // Wait for the port-forward proxy to be ready.
+                p.report({ message: "Waiting for port-forwarding ready..."});
+                kubeChannel.showOutput("Starting to wait the port proxy to be ready...", "Wait for port-forward");
+                await this.waitForProxyReady(proxyResult.proxyProcess);
+                kubeChannel.showOutput("Finished to wait.");
+                
+                // Start debug session.
+                p.report({ message: `Starting ${this.debugProvider.getDebuggerType()} debug session...`});
+                kubeChannel.showOutput("Starting to start debug session...", "Start debug");
+                const sessionName = appName || `${Date.now()}`;
+                await this.startDebugging(cwd, sessionName, proxyResult.proxyDebugPort, proxyResult.proxyAppPort, async () => {
+                    proxyResult.proxyProcess.kill();
+                    if (appName) {
+                        await this.cleanupResource(`deployment/${appName}`);
+                    }
+                });
+            } catch (error) {
+                vscode.window.showErrorMessage(error);
+                kubeChannel.showOutput(`Debug(Launch) on kubernetes failed. See the errors: ${error}.`);
+                if (appName) {
+                    await this.cleanupResource(`deployment/${appName}`);
+                }
+            }
+            return null;
+        });
+    }
+
+    /**
+     * In attach mode, the user should choose the running pod first, then the debugger will attach to it.
+     * 
+     * @param workspaceFolder the active workspace folder.
+     * @param pod the debug pod name.
+     */
+    public async attachDebug(workspaceFolder: vscode.WorkspaceFolder, pod?: string): Promise<void> {
+        // TODO
+        return;
+    }
+
+    private async createPortForward(podName, debugPort, appPort): Promise<any> {
+        const portMapping = [];
+        const portfinder = require('portfinder');
+        // Find a free local port for forwarding data to remote app port.
+        let proxyAppPort = 0;
+        if (appPort) {
+            proxyAppPort = await portfinder.getPortPromise({
+                port: appPort
+            });
+            portMapping.push(proxyAppPort + ":" + appPort);
+        }
+        // Find a free local port for forwarding data to remote debug port.
+        const proxyDebugPort = await portfinder.getPortPromise({
+            port: Math.max(10000, Number(proxyAppPort) + 1)

```

Comment:
You'are right. `portfinder` doesn't guarantees `proxyAppPort` is occupied, that's why i start searching from `proxyAppPort + 1` to avoid finding the same free port.

`portfinder` will start searching from the specified base port.

Reply:
Bah, I misread - you are right, ignore me, sorry!